src/expr.rs

```diff
@@ -19,194 +19,246 @@ use syntax::print::pprust;
+    let mut result = String::with_capacity(round_up_to_power_of_two(s.len()));
+    result.push('"');
+    loop {
+        let max_chars = if cur_start == 0 {
+            // First line.
+            width - 2 // 2 = " + \
+        } else {
+            config!(max_width) - offset - 1 // 1 = either \ or ;
+        };
 
-            // Make sure we're on a char boundary.
-            cur_end = next_char(&s, cur_end);
+        let mut cur_end = cur_start + max_chars;
 
-            // Push cur_end left until we reach whitespace
-            while !s.char_at(cur_end-1).is_whitespace() {
-                cur_end = prev_char(&s, cur_end);
+        if cur_end >= s.len() {
+            result.push_str(&s[cur_start..]);
+            break;
+        }
 
-                if cur_end - cur_start < MIN_STRING {
-                    // We can't break at whitespace, fall back to splitting
-                    // anywhere that doesn't break an escape sequence
-                    cur_end = next_char(&s, cur_start + max_chars);
-                    while s.char_at(prev_char(&s, cur_end)) == '\\' {
-                        cur_end = prev_char(&s, cur_end);
-                    }
-                    break;
+        // Make sure we're on a char boundary.
+        cur_end = next_char(&s, cur_end);
+
+        // Push cur_end left until we reach whitespace
+        while !s.char_at(cur_end-1).is_whitespace() {
+            cur_end = prev_char(&s, cur_end);
+
+            if cur_end - cur_start < MIN_STRING {
+                // We can't break at whitespace, fall back to splitting
+                // anywhere that doesn't break an escape sequence
+                cur_end = next_char(&s, cur_start + max_chars);
+                while s.char_at(prev_char(&s, cur_end)) == '\\' {
+                    cur_end = prev_char(&s, cur_end);
                 }
+                break;
             }
-            // Make sure there is no whitespace to the right of the break.
-            while cur_end < s.len() && s.char_at(cur_end).is_whitespace() {
-                cur_end = next_char(&s, cur_end+1);
-            }
-            result.push_str(&s[cur_start..cur_end]);
-            result.push_str("\\\n");
-            result.push_str(indent);
-
-            cur_start = cur_end;
         }
-        result.push('"');
+        // Make sure there is no whitespace to the right of the break.
+        while cur_end < s.len() && s.char_at(cur_end).is_whitespace() {
+            cur_end = next_char(&s, cur_end+1);
+        }
+        result.push_str(&s[cur_start..cur_end]);
+        result.push_str("\\\n");
+        result.push_str(indent);
 
-        result
+        cur_start = cur_end;
     }
+    result.push('"');
 
-    fn rewrite_call(&mut self,
+    Some(result)
+}
+
+fn rewrite_call(context: &RewriteContext,
                     callee: &ast::Expr,
                     args: &[ptr::P<ast::Expr>],
                     width: usize,
                     offset: usize)
```

Comment:
Please fix the indentation here (this should make us fail a test, I wonder why travis passes?)